### PR TITLE
MBL-798: SetPasswordViewModel.kt to Jetpack ViewModel and RxJava2.2

### PR DESCRIPTION
**Pull request by leighdouglas**
# 📲 What

Migration to rxjava 2.2 and jetpacks viewmodel

# 🤔 Why

[Technical Details](https://kickstarter.atlassian.net/wiki/spaces/NT/pages/2350317579/ViewModel+Tech+Debt+Epic#Technical-details)

# 👀 See

No user facing changes

# 📋 QA

This feature is tricky to test - I had to manually set the `needsPassword` field in the User struct to `true`. I then had to make sure the feature flag for facebook login deprecation was on or the condition was commented out in the discovery view model. This is what made the set password screen appear after logging in.

On the actual set password screen: 
Make sure that the warning text changes if the passwords in each field do not match, or if the password in the first field is fewer than 6 characters. 

To test success or failure in setting the password, some work is required from the backend if you want to QA this

# Story 📖

[MBL-766: FrequentlyAskedQuestionsViewHolderViewModel.kt to Jetpack ViewModel and RxJava2.2](https://kickstarter.atlassian.net/browse/MBL-766)

### CODE REVIEW

File: app/src/main/java/com/kickstarter/viewmodels/SetPasswordViewModel.kt [hunk 1]
```diff
@@ -93,65 +99,69 @@ interface SetPasswordViewModel {
 
             setNewPassword
                 .map { it.warning() }
+                .filter { ObjectUtils.isNotNull(it) }
+                .map { requireNotNull(it) }
+                .filter { it != 0 }
```
Comment (leighdouglas) [inline app/src/main/java/com/kickstarter/viewmodels/SetPasswordViewModel.kt:104]: I had to update the function `warning()` to return a non-null value, as it was still causing the app to crash due to a the mapper function returning a null value, even with the `isNotNull` checks in the chain. I'm not exactly sure why this is the case here or what makes this different, as opposed to other places where a null check in the chain is enough to not cause a crash (the intent rx chain above is an example of this, where the getStringExtra returns a null value if there is no extra with that key attached to the intent, but it does not crash.)

File: app/src/main/java/com/kickstarter/ui/activities/SetPasswordActivity.kt [hunk 2]
```diff
@@ -34,51 +44,52 @@ class SetPasswordActivity : BaseActivity<SetPasswordViewModel.ViewModel>() {
...
 
         this.viewModel.outputs.setUserEmail()
-            .compose(bindToLifecycle())
-            .compose(Transformers.observeForUI())
+            .observeOn(AndroidSchedulers.mainThread())
             .subscribe {
-                binding.setPasswordHint.text = this.environment().ksString()?.format(getString(R.string.We_will_be_discontinuing_the_ability_to_log_in_via_FB), "email", it)
-            }
+                binding.setPasswordHint.text = getEnvironment()?.ksString()?.format(getString(R.string.We_will_be_discontinuing_the_ability_to_log_in_via_FB), "email", it)
+            }.addToDisposable(disposables)
```
Comment (mtgriego) [inline app/src/main/java/com/kickstarter/ui/activities/SetPasswordActivity.kt:56]: i know this was in the last iteration of this as well, but do we have a translated string for "email"?

**Review by mtgriego — APPROVED** [review]

File: app/src/main/java/com/kickstarter/ui/activities/SetPasswordActivity.kt [hunk 3]
```diff
@@ -88,8 +99,4 @@ class SetPasswordActivity : BaseActivity<SetPasswordViewModel.ViewModel>() {
     private fun setFormDisabled(isDisabled: Boolean) {
         setFormEnabled(!isDisabled)
     }
-
-    override fun back() {
```
Comment (Arkariang) [inline app/src/main/java/com/kickstarter/ui/activities/SetPasswordActivity.kt:92]: @leighdouglas I think the disable back action was on purpose, to not allow the user to leave the screen

**Review by Arkariang — APPROVED** [review]

File: app/src/main/java/com/kickstarter/ui/activities/SetPasswordActivity.kt (hunk 3 above)
Comment (leighdouglas) [reply app/src/main/java/com/kickstarter/ui/activities/SetPasswordActivity.kt:92]: Good catch, there is no back button, but a user can swipe or press the backbutton to exit. I'll bring this up during sprint planning and add this as a QA script.

File: app/src/main/java/com/kickstarter/ui/activities/SetPasswordActivity.kt (hunk 2 above)
Comment (leighdouglas) [reply app/src/main/java/com/kickstarter/ui/activities/SetPasswordActivity.kt:56]: We do! i can update in a subsequent PR. Thanks for catching this.

**Pull request merged**